### ISSUE TITLE
Questions on permissions

### ISSUE THREAD

**Issue by IzzySoft** [warren-bank/Android-AirPlay-Client#9]
Your latest release threw up some warnings here:

```
! repo/com.github.warren_bank.airplay_client_50614.apk declares flag(s): usesCleartextTraffic
! repo/com.github.warren_bank.airplay_client_50614.apk declares sensitive permission(s):
  android.permission.MANAGE_EXTERNAL_STORAGE android.permission.READ_EXTERNAL_STORAGE
  android.permission.READ_MEDIA_AUDIO android.permission.READ_MEDIA_IMAGES
  android.permission.READ_MEDIA_VIDEO
! repo/com.github.warren_bank.airplay_client_50614.apk contains signature block blobs: 0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)
```

I guess `usesCleartextTraffic` is for local connections which do not support certificates, so I can add it to the "green list" with that explanation if you confirm. But what to put for the list of permissions?

As for `DEPENDENCY_INFO_BLOCK`, that can easily be avoided with a small addition to your `build.gradle`:

```gradle
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains. More details can be found e.g. here: [Ramping up security: additional APK checks are in place with the IzzyOnDroid repo](https://android.izzysoft.de/articles/named/iod-scan-apkchecks#blobs).

Thanks in advance!

**Comment by warren-bank** [comment]
This app allows the user to browse the local device for audio and video files. Then starts a local (http) web server. Then sends a (http) request to a video player that includes the URL for the selected media file, which can be accessed on the LAN.

* usesCleartextTraffic
  - all (http) network traffic
* READ_EXTERNAL_STORAGE, READ_MEDIA_IMAGES, READ_MEDIA_VIDEO
  - to find and access local audio and video files
* DEPENDENCY_INFO_BLOCK
  - I wasn't aware of [this](https://android.izzysoft.de/articles/named/iod-scan-apkchecks) and will update my _build.gradle_ in all of my apps.. the next time that I touch each

**Comment by warren-bank** [comment]
As a comment to casual users who might be seeing this issue and are now concerned about the most recent release..

Nothing has changed in the app. The most recent update was a _very minor_ tweak to address an issue raised by a user.

This notification is a courtesy by _Izzy_ that pertains to some security checks that have recently been added to his f-droid repo.

**Comment by IzzySoft** [comment]
Thanks Warren! Updated the metadata (effective immediately) for the ones you've listed. Only one remaining would be `android.permission.MANAGE_EXTERNAL_STORAGE` then. Guess something with SAF trouble somewhere between Android 5 and 9?

**Comment by warren-bank** [comment]
oh, sorry.. my summary was incomplete.. I copied one permission per line.. didn't notice they weren't formatted that way

correction:
* _READ_EXTERNAL_STORAGE_, _READ_MEDIA_AUDIO_, _READ_MEDIA_VIDEO_, _READ_MEDIA_IMAGES_, _MANAGE_EXTERNAL_STORAGE_
  - to find and access local audio, video, and image files

off the top of my head..
* _READ_EXTERNAL_STORAGE_ worked perfectly <= Android 9
* Android 10 broke everything..
  - _READ_MEDIA_AUDIO_, _READ_MEDIA_VIDEO_, _READ_MEDIA_IMAGES_ is what was introduced.. but didn't work at all
  - I should've added to the manifest: `<application android:requestLegacyExternalStorage="true" ... />`,<br>but apparently didn't.. and consequently this app probably won't work too well on A10
* Android 11 introduced _MANAGE_EXTERNAL_STORAGE_, which allows me to use the filesystem as I did pre- A10

so.. yeah.. to summarize.. the app doesn't ask for any permissions that it doesn't use..<br>unless you count _READ_MEDIA_xxx_ which I find to be pretty useless.. but (at least in theory) are needed.

**Comment by IzzySoft** [comment]
Thanks again! My humble knowledge concerning those:

* `READ_MEDIA_*` are intended as addition to SAF to protect files which might be considered rather personal and thus needing extra protection. I understand the iddea of protecting those files, of course – though I do _not_ understand it in context with SAF, as there one has to explicitly pick the location anyway. Surely we are both lacking some wisdom there to fully grasp the ide :see_no_evil:
* `MANAGE_EXTERNAL_STORAGE` is rather intended for file managers and such. Though with SAF, on some Android versions at least that is needed if you want to access locations like `Downloads/` or the sd card root directory. So shall I put that up as description there: "to allow accessing files to share from all storage locations"?

> to summarize.. the app doesn't ask for any permissions that it doesn't use.

I never thought you would :stuck_out_tongue_winking_eye: This is rather about transparency – so those browsing the catalog and wonder can have an explanation and better understand. We both know the Android permissions are something like "sealed books" to the less tech-savvy (and sometimes even to us)…

**Comment by warren-bank** [comment]
I agree.. I'm certainly no expert in the new SAF. I thought that the old "legacy" methodology made perfect sense. Maybe there is a better way to request permission to read from and write to the filesystem on A10+, but I haven't been able to find it. It seems to me that `MANAGE_EXTERNAL_STORAGE` is the only permission that allows me to display a file/directory picker dialog.. that can navigate up/down through the directory tree. Basic functionality.. I would say. But yes.. you're right that the documentation calls it _dangerous_.. and _Google Play_ would probably disallow it.

To be perfectly honest, I still think that Android __5__ was the best version of the OS.. and they should've just stopped there. Every version since has only removed features and functionality.. in the name of improved security. &lt;/rant&gt;

**Comment by IzzySoft** [comment]
> I thought that the old "legacy" methodology made perfect sense

Depends on what for. The new SAF has its advantages, e.g. allowing an app just access to a single directory and not the entire filesystem if you just want an option for export/import. But it also has its disadvantages, e.g. when it comes to speed (think of apps to sync entire directory trees for example).

> It seems to me that `MANAGE_EXTERNAL_STORAGE` is the only permission that allows me to display a file/directory picker dialog

That is weird. The entire SAF concept bases on the fact that the app pops up a file picker to choose the directory to give access to. And there are many apps using it this way, without having requested `MANAGE_EXTERNAL_STORAGE`. A quick search brought up [Open files using the Storage Access Framework](https://developer.android.com/guide/topics/providers/document-provider) for me, which states SAF being introduced with Kitkat (and there I thought it was Lollipop) and how the workflow goes, including a "Document Picker". Maybe the example code [here](https://stackoverflow.com/q/58899075/2533433) can shed some light? Or [this article](https://www.holdapp.com/blog/how-to-access-files-in-android-storage-access-framework-and-activity-result-api)? I'm not an Android dev myself, so I cannot really tell. Also see the official documentation on [Access documents and other files from shared storage](https://developer.android.com/training/data-storage/shared/documents-files?hl=en) maybe:

> The [`ACTION_OPEN_DOCUMENT_TREE`](https://developer.android.com/reference/android/content/Intent#ACTION_OPEN_DOCUMENT_TREE) intent action, available on Android 5.0 (API level 21) and higher, allows users to select a specific directory, granting your app access to all of the files and sub-directories within that directory.

But that of course would only cover Android 5 and above. Before that, SAF wasn't really usable, if available at all.

> and _Google Play_ would probably disallow it.

I'm pretty sure they would, yeah.

> Every version since has only removed features and functionality.. in the name of improved security.

Not starting with LP but much earlier. My first Android was Eclair (2.1), and removal of functionality for me already started with Froyo (2.2)…

**Comment by warren-bank** [comment]
> Not starting with LP but much earlier.

Totally agree. The only reason that I stick a pin at v5 (LP) is that it adds a much needed feature.. the upgradeable system WebView. I _wish_ that was available in v4 (ICS and KitKat).

I suppose you could debate that runtime permissions (v6) and split screen (v7) are worth-while additions to the OS. As a user, runtime permissions are nice.. as a developer, they're a PITA. As a user, I've personally never wanted a split screen.

As for file/directory pickers.. I _suppose_ using an Intent to offload the behavior to an external app could work. I just like including my own (minimal) view of the filesystem to allow the user to do whatever fs task is needed. That _was_ a very common pattern in older apps. Now, I suppose, it's considered an anti-pattern?

results for [quick search](https://www.google.com/search?q=android+file+picker+scoped+github):
* [related issue](https://github.com/DroidNinja/Android-FilePicker/issues/305) for a file picker library
* [demo app](https://github.com/yogeshpaliyal/Android-File-Picker-With-Scoped-Storage) for using Intents

**Comment by IzzySoft** [comment]
TBH, I cannot remember ever having used split screen :see_no_evil:

As for the pickers: I wouldn't exactly call it an anti-feature. IMHO it's rather a matter of trust here: do I trust the app to give it access to everything on my storage? Do I trust its developer, whom I might not even know? And apart from trust, mistakes happen to the best of us. So some part of isolation is… neat.

Which is why I usually don't press at this point. I notify, yes. I try to find out, together with the authors, if a different approach would make more sense. If it's just for export/import of app settings – OK, there I push a little, because for that SAF makes perfect sense. For an app like Airplay Client, full storage access _might_ make sense – e.g. if sharing from every location is desirable, which I can imagine. And that's where I find the SAF limitation nuts: if someone wants to give an app access to Downloads or the entire SD card, why not simply respect that choice?

And there we're back to the other problem: that at Android, with each new carrot comes a stick. And one cannot help but, in terms of storage & Co, to ask: "Ein Schelm, wer Böses dabei denkt!" (roughly translated: "a rascal who thinks of evil intention here"). Backups only work to Google Servers. Storage is easier accessed in the Cloud. And so on. But I guess I'm preaching the choir here…

TL;DR: If you're going to stick with those permissions, just let me know what explanation I should attach and I'll be fine.

**Comment by warren-bank** [comment]
earlier, without giving it any thought, my explanation was: `to find and access local audio, video, and image files`

that summary works for me.. it's simple and to the point

**Comment by IzzySoft** [comment]
Thanks! Added that for MANAGE then: "to find and access audio, video, and image files everywhere on local storage".

All done then I'd say – feel free to close if you no longer need it open as a reminder for something. And thanks again!
